### PR TITLE
Add RCW for IDataObject

### LINKED ISSUE #7518: .NET 7.0 preview DragEnter bug after update to latest versions.
#### When trying to drag and drop files from folders or other sources into a Windows.Forms.Controls:

NotImplementedException in [IDropTargetVtbl.cs](https://source.dot.net/#System.Windows.Forms.Primitives/Interop/WinFormsComWrappers.IDropTargetVtbl.cs,77)
VS2022 preview, .NET 7.0 preview, SDK, all latest versions.

It works compiled for  .NET 6.0, it worked for all .NET 7.0 previews before.
Example project is available at [GitHub](https://github.com/c-ohle/RationalNumerics)

image

### Reproduction Steps

Should be easy to reproduce: 
VS2022 - New Project - Windows Forms App - .NET 7.0 preview
Form1 - AllowDrop = true; - override OnDragEnter
on file drop -> exception

### Expected behavior

Should work or message/exception if there is a security problem or what ever. 

### Actual behavior

NotImplementedException

### Regression?

_No response_

### Known Workarounds

no

### Configuration

Processor: Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz (2112 MHz) × 8
System: Microsoft Windows 10.0.22622 X64 Runtime: .NET 7.0.0-preview.6.22324.4 X64
BigRational: 1.0.0.5 Runtime: v4.0.30319 .NETCoreApp,Version=v7.0 

For Test project on: [GitHub sample](https://github.com/c-ohle/RationalNumerics)

### Other information

_No response_

### PR DESCRIPTION
**Pull request by kant2002**
Fixes #7518

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7519)

### CODE REVIEW

**Comment by kant2002** [comment]
Or better, I should just finish https://github.com/dotnet/winforms/pull/7087

**Comment by RussKie** [comment]
Thank you. However, the biggest qestion here is how we test this? And ensure we haven't missed anything else...

> Or better, I should just finish #7087

We're in RC1 and the runway is too small for any significant changes, and risks are too great

**Comment by willibrandon** [comment]
I can help test this, possibly. It needs to be an end to end test like the drag & drop tests in `UIIntegration`, but the drag operation needs to start from Explorer. My problem is I don't know how to automate the mouse drag & drop operation from Explorer yet.

If I knew how to get the coordinates of a file located on the screen, I could write the test.

**Comment by kant2002** [comment]
I thin we need something more complicated then `UIIntegration`. Maybe some "test running" which will run EXE files and then perform interaction over the OS. Maybe we can automate interaction using https://github.com/Microsoft/WinAppDriver . Probably that way we will test and Accessibility itself albeit indirectly.

**Comment by kant2002** [comment]
Basically less risky change would be undo these 2 changes in this file https://github.com/dotnet/winforms/pull/6976/files#diff-efdc83ac4513276157401c19a46f9c20f7c2aa84d9b9eb1cd19f37aca86420d5R33
Also we should add tests, so this issue will never appears on our radar.

**Comment by RussKie** [comment]
> I can help test this, possibly. It needs to be an end to end test like the drag & drop tests in `UIIntegration`, but the drag operation needs to start from Explorer. My problem is I don't know how to automate the mouse drag & drop operation from Explorer yet.

It's possible to open an new instance of Explorer and select a file in it, e.g. `cmd /c "explorer /select,<full path to file>"`. More docs are available at https://ss64.com/nt/explorer.html.
The question is how we can find the screen coordinates of that file... WinAppDriver appears to able to do this somehow: https://github.com/microsoft/WinAppDriver/blob/9c561a52ace7b57bb375bc752389731a68dbcc53/Samples/C%23/NotepadTest/ScenarioPopupDialog.cs#L81.

**Comment by RussKie** [comment]
> I can help test this, possibly. It needs to be an end to end test like the drag & drop tests in `UIIntegration`, but the drag operation needs to start from Explorer. My problem is I don't know how to automate the mouse drag & drop operation from Explorer yet.

It's possible to open an new instance of Explorer and select a file in it, e.g. `cmd /c "explorer /select,<full path to file>"`. More docs are available at https://ss64.com/nt/explorer.html.
The question is how we can find the screen coordinates of that file... ~WinAppDriver appears to able to do this somehow: https://github.com/microsoft/WinAppDriver/blob/9c561a52ace7b57bb375bc752389731a68dbcc53/Samples/C%23/NotepadTest/ScenarioPopupDialog.cs#L81.~

Files in Windows Explorer are displayed in a ListView control, so each file/folder is just a listview item. Accessibility tools (e.g., Narrator, etc.) can work it out, so we should be able too.
image

Perhaps this could help https://docs.microsoft.com/windows/win32/winauto/uiauto-clientportal and https://docs.microsoft.com/windows/win32/api/uiautomationclient/nf-uiautomationclient-iuiautomation-getfocusedelement

Pinging @vladimir-krestov for ideas.

**Comment by willibrandon** [comment]
> It's possible to open an new instance of Explorer and select a file in it, e.g. cmd /c "explorer /select,<full path to file>". More docs are available at https://ss64.com/nt/explorer.html.

Thank you.

I'm still trying to answer the question of how we can reliably find the screen coordinates of that file, but for the record, I can  reproduce the issue in a `UIIntegration` test if I hard code the Explorer screen coordinates and drag to the application.

NotImplementedException

**Comment by willibrandon** [comment]
A manual test that already exists and reproduces this issue is in `WinformsControlsTest.DragDrop`.

**Comment by willibrandon** [comment]
I will test these changes.

File: src/System.Windows.Forms.Primitives/src/Interop/WinFormsComWrappers.DataObjectWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            }
+
+            public void GetData(ref FORMATETC format, out STGMEDIUM medium)
+            {
+                fixed (FORMATETC* pFormat = &format)
+                {
+                    IDataObjectVtbl.STGMEDIUM_Raw mediumRaw;
+                    ((delegate* unmanaged<IntPtr, FORMATETC*, IDataObjectVtbl.STGMEDIUM_Raw*, HRESULT>)(*(*(void***)_wrappedInstance + 3)))
+                        (_wrappedInstance, pFormat, &mediumRaw).ThrowIfFailed();
+                    medium.pUnkForRelease = Marshal.GetObjectForIUnknown(mediumRaw.pUnkForRelease);
```
Comment (willibrandon) [inline src/System.Windows.Forms.Primitives/src/Interop/WinFormsComWrappers.DataObjectWrapper.cs:37]: What if `pUnk` is `Zero`?

**Comment by merriemcgaw** [comment]
> Basically less risky change would be undo these 2 changes in this file https://github.com/dotnet/winforms/pull/6976/files#diff-efdc83ac4513276157401c19a46f9c20f7c2aa84d9b9eb1cd19f37aca86420d5R33 Also we should add tests, so this issue will never appears on our radar.

@kant2002 the team has chatted and we think it's probably for the best that we do as you suggest here for .NET 7, and we take the completed fix in .NET 8 with the additional testing. Does that work for you?

**Comment by kant2002** [comment]
Yes. I think so too. I think smallest thing to undo would be just 2 lines but if you think whole PR should be undone I'm fine. We should have stable WinForms a priority

**Comment by dreddy-work** [comment]
> Yes. I think so too. I think smallest thing to undo would be just 2 lines but if you think whole PR should be undone I'm fine. We should have stable WinForms a priority

if we do not have any other known scenario regressed and this reported scenario is fixed by undoing those two lines, we should stick to those lines in this and monitor for any new reports. Meanwhile, I will ask our test team to run some drag/drop scenarios to see if there are any pri0 scenarios that need to be addressed?

**Comment by Olina-Zhang** [comment]
@kant2002 tested this PR using 2 sample projects, the issue still repro. Please take a look, thanks.
One project is from https://github.com/c-ohle/RationalNumerics with targeting the latest .Net 7.0
drag-drop
Another is the one we created about drag-dropping image on pictureBox in Winforms from file explorer:
[WinFormsApp14.zip](https://github.com/dotnet/winforms/files/9256025/WinFormsApp14.zip)
drag-drop_NotFixed

**Comment by dreddy-work** [comment]
@kant2002 , did you get a chance to look latest comments here?

**Comment by RussKie** [comment]
> Basically less risky change would be undo these 2 changes in this file [#6976 (files)](https://github.com/dotnet/winforms/pull/6976/files#diff-efdc83ac4513276157401c19a46f9c20f7c2aa84d9b9eb1cd19f37aca86420d5R33)

Would you be able to action this? The window for RC1 is closing fast.

**Comment by kant2002** [comment]
If only undo, I can do that.

**Comment by dreddy-work** [comment]
@kant2002 , would it be possible to make the requested changes here before coming Monday?

**Comment by willibrandon** [comment]
@RussKie - I now know how to write the automated test, dragging from Explorer, using `IUIAutomation::GetFocusedElement` and `IUIAutomationElement::GetClickablePoint`. Your help has been invaluable.

**Comment by RussKie** [comment]
> @RussKie Igor Velikorossov FTE - I now know how to write the automated test, dragging from Explorer, using [`IUIAutomation::GetFocusedElement`](https://docs.microsoft.com/en-us/windows/win32/api/uiautomationclient/nf-uiautomationclient-iuiautomation-getfocusedelement) and [`IUIAutomationElement::GetClickablePoint`](https://docs.microsoft.com/en-us/windows/win32/api/uiautomationclient/nf-uiautomationclient-iuiautomationelement-getclickablepoint). Your help has been invaluable.

You're giving me too much credit. I randomly blurbed something and you did all the work :) And now I'm eager to learn how to do that, you'll have to show us a test.

**Comment by willibrandon** [comment]
> You're giving me too much credit. I randomly blurbed something and you did all the work :) And now I'm eager to learn how to do that, you'll have to show us a test.

I mean all of your help so far, you are like a code whisperer.

I'm unsure how to proceed with the test, should I create a separate issue?

**Comment by RussKie** [comment]
> I'm unsure how to proceed with the test, should I create a separate issue?

Yes please.

**Pull request closed without merging**